Adding has year

jessicant opened the pull request:
#### Check List:
* Fixes part of #2541
* Unit tests : YES 
* Javadoc with a code example (on API only) : YES
* PR meets the [contributing guidelines](https://github.com/assertj/assertj-core/blob/main/CONTRIBUTING.md)

Following the [contributing guidelines](https://github.com/assertj/assertj-core/blob/main/CONTRIBUTING.md) will make it easier for us to review and accept your PR.

onacit commented:
This is a little bit different what I thought.

`hasYear(I)` should to be defined in `AbstractLocalDateAssert` as `hasHour(I)` should be defined `AbstractLocalTimeAssert`.

And I believe that `LocalDate` and `LocalTime` can be extracted from an assertion of `LocalDateTimeAssert`.

```java
final LocalDate localDate = LocalDate.now();
assertThat(localDate).hasYear(actual.getYear();

final LocalTime localTime = LocalTime.now();
assertThat(localTime).hasHour(localTime.getHour());

final LocalDateTime localDateTime = localDate.atTime(localTime);
assertThat(localDateTime)
    .extracting(a -> assertThat(a.toLocalDate()))
    .hasMonth(localDate.getMonth());
assertThat(localDateTime)
    .extracting(a -> assertThat(a.toLocalTime()))
    .hasMinute(localTime.getMinute());
```

jessicant commented:
Ah, I see. Apologies, I am a student and this is my first time working with a larger codebase. 

I will correct this.

onacit commented:
@jessicant You don't have to care about `extractingLocal(Date|Time)` part. Thanks.

joel-costigliola commented on src/main/java/org/assertj/core/error/ShouldEqualYear.java:
```diff
@@ -0,0 +1,52 @@
+ *
+ * @author Jessica Hamilton
+ */
+public class ShouldEqualYear extends BasicErrorMessageFactory {
```
To be consistent with date assertions, let's reuse `ShouldHaveDateField`, In `Dates.assertHasYear` it is used like `shouldHaveDateField(actual, "year", year)`:
```java
  public void assertHasYear(AssertionInfo info, Date actual, int year) {
    assertNotNull(info, actual);
    if (yearOf(actual) != year) throw failures.failure(info, shouldHaveDateField(actual, "year", year));
  }
```

We can remove `ShouldEqualYear` from this PR then.

jessicant replied:
This makes sense, I removed it in a later commit.

joel-costigliola commented:
@jessicant you don't need to close the PR when you are working on it, leave it open and when it's ready for review let us know or simply keep your commits local and push them when they are ready to be reviewed

joel-costigliola commented on src/test/java/org/assertj/core/api/localdatetime/LocalDateTimeAssert_hasYear_Test.java:
```diff
@@ -0,0 +1,61 @@
+
+  @Test
+  void should_pass_if_actual_is_in_year_2022() {
+    assertThat(refLocalDateTime).hasYear(2022);
```
I don't think this test is going to succeed in 2023 as `refLocalDateTime` is set to `now()` ;-)
Instead initialize it with `LocalDateTime.of()` where you can set explicitly the year, month, ... values.

while we are at it, rename `refLocalDateTime` to `actual` or `actualLocalDateTime` (use wichever you like best)

joel-costigliola commented on src/test/java/org/assertj/core/api/localdatetime/LocalDateTimeAssert_hasYear_Test.java:
```diff
@@ -0,0 +1,61 @@
+    AssertionError assertionError = expectAssertionError(() -> assertThat(refLocalDateTime).hasYear(2021));
+    // THEN
+    then(assertionError).hasMessage(format("%nExpecting year of:%n  " + refLocalDateTime +
+      " (java.time.LocalDateTime)%nto be year:%n  2021%n"));
```
replace the string error message by a call to the ErrorMessageFactory as in 
```java
assertThatAssertionErrorIsThrownBy(code).withMessage(shouldBeBefore(AFTER, REFERENCE).create());
```
the rationale being that if the error message wording changes this test should not be impacted.

in your case, it should be (assuming you are using:
```java
    // GIVEN
    int expectedYear = 2021;
    // WHEN
    AssertionError assertionError = expectAssertionError(() -> assertThat(refLocalDateTime).hasYear(expectedYear));
    // THEN
    then(assertionError).hasMessage(shouldHaveDateField(refLocalDateTime, "year", expectedYear).create());
```

joel-costigliola commented on src/test/java/org/assertj/core/api/localdatetime/LocalDateTimeAssert_hasYear_Test.java:
```diff
@@ -0,0 +1,61 @@
+  void should_throw_error_if_given_localdatetime_is_null() {
+    assertThatIllegalArgumentException().isThrownBy(() -> assertThat(refLocalDateTime).isEqualToIgnoringHours(null))
+      .withMessage(NULL_LOCAL_DATE_TIME_PARAMETER_MESSAGE);
+  }
```
this test is not relevant here since it's not calling `hasYear`  (likely some copy paste), let's remove it

joel-costigliola commented on src/test/java/org/assertj/core/api/localdatetime/LocalDateTimeAssert_hasYear_Test.java:
```diff
@@ -0,0 +1,61 @@
+    // GIVEN
+    LocalDateTime actual = null;
+    // WHEN
+    AssertionError assertionError = expectAssertionError(() -> assertThat(actual).isEqualToIgnoringHours(LocalDateTime.now()));
```
this should call  `hasYear`

jessicant commented:
@joel-costigliola Thank you for the feedback! That's all very helpful, I will rework this during this week

jessicant commented:
@onacit I have reworked it into AbstractLocalDateAssert, let me know if this is still not what you were thinking.

@joel-costigliola I believe I have addressed most of your comments. Kindly let me know your thoughts on the new commits when you get the chance place :)

onacit commented:
@jessicant What a great work you've done! I Thank you.

joel-costigliola commented on src/main/java/org/assertj/core/util/DateUtil.java:
```diff
@@ -22,6 +22,7 @@
 import java.text.DateFormat;
 import java.text.ParseException;
 import java.text.SimpleDateFormat;
+import java.time.LocalDate;
```
I don't see it being used, if that's true remove this import

joel-costigliola replied:
I think there is no need to change this file for this PR, let's revert all the changes as they are not relevant to the PR

joel-costigliola commented on src/test/java/org/assertj/core/api/localdate/LocalDateAssert_hasYear_Test.java:
```diff
@@ -0,0 +1,59 @@
+import org.junit.jupiter.api.Test;
+
+
+public class LocalDateAssert_hasYear_Test {
```
class does not need to be public, otherwise nice tests!

joel-costigliola commented on src/test/java/org/assertj/core/api/localdate/LocalDateAssert_hasYear_Test.java:
```diff
@@ -0,0 +1,59 @@
+//  void should_throw_error_if_given_localdate_is_null() {
+//    assertThatIllegalArgumentException().isThrownBy(() -> assertThat(refLocalDate).isEqualToIgnoringHours(null))
+//      .withMessage(NULL_LOCAL_DATE_TIME_PARAMETER_MESSAGE);
+//  }
```
remove this unrelevant commented test

joel-costigliola commented on src/main/java/org/assertj/core/api/AbstractLocalDateTimeAssert.java:
```diff
@@ -817,4 +818,11 @@ private static boolean haveSameYearAndMonth(LocalDateTime actual, LocalDateTime
+   * @param actual the actual datetime. expected not be null
+   * @param year the given year. expected not be null
+   * @return true if the datetime year is equal to the given year, false otherwise
+   */
```
at some point we will want to add `public SELF hasYear(int year)` for `LocalDateTime`, you can add it now if you feel like it or that could be done in another by you or someone else, if you don't no problem just revert the changes in this file in this case

jessicant replied:
I will add it in another PR unless someone gets to it first.

joel-costigliola commented:
@jessicant a bit of code clean up and we will be able to integrate this PR, thanks for the work.

jessicant commented:
@joel-costigliola I think I addressed everything in your comments and got the code tidied up. Let me know if there is anything else I need to fix.

joel-costigliola commented:
Integrated thanks @jessicant!
